### PR TITLE
Optimize Arithmetic Circuits

### PR DESCRIPTION
**Pull request by hovanja2011**

### CODE REVIEW

File: symbolic-base/src/ZkFold/Base/Algebra/Basic/Class.hs [hunk 1]
```diff
@@ -690,3 +690,76 @@ instance FromConstant b a => FromConstant b (p -> a) where
...
+
+instance (AdditiveSemigroup a) => AdditiveSemigroup (Maybe a) where
+    (+) :: Maybe a -> Maybe a -> Maybe a
+    (+) = liftA2 (+)
+
+instance (MultiplicativeSemigroup a) => MultiplicativeSemigroup (Maybe a) where
+    (*) :: Maybe a -> Maybe a -> Maybe a
+    (*) = liftA2 (*)
+
+instance (FromConstant Natural a, MultiplicativeSemigroup a) => Scale Natural (Maybe a)
```
Comment (TurtlePU) [inline symbolic-base/src/ZkFold/Base/Algebra/Basic/Class.hs:713]: Better to use the `Scala Natural a` instance via `fmap`

File: symbolic-base/src/ZkFold/Base/Algebra/Basic/Class.hs [hunk 2]
```diff
@@ -690,3 +690,76 @@ instance FromConstant b a => FromConstant b (p -> a) where
...
+instance (FromConstant Integer a, MultiplicativeSemigroup a) => Scale Integer (Maybe a)
```
Comment (TurtlePU) [inline symbolic-base/src/ZkFold/Base/Algebra/Basic/Class.hs:714]: Better to use the `Scale Integer a` instance via `fmap`

File: symbolic-base/src/ZkFold/Base/Algebra/Basic/Class.hs [hunk 3]
```diff
@@ -690,3 +690,76 @@ instance FromConstant b a => FromConstant b (p -> a) where
...
+instance (MultiplicativeMonoid a, Exponent a Natural) => MultiplicativeMonoid (Maybe a) where
+    one :: Maybe a
+    one = Just one
+
+instance (AdditiveMonoid a, MultiplicativeMonoid a, FromConstant Natural a) => Semiring (Maybe a)
+
+instance (AdditiveMonoid a, Scale Integer a, FromConstant Integer a, MultiplicativeSemigroup a, FromConstant Natural a) => AdditiveGroup (Maybe a) where
+
+    (-) :: Maybe a -> Maybe a -> Maybe a
+    x - y = x + negate y
```
Comment (TurtlePU) [inline symbolic-base/src/ZkFold/Base/Algebra/Basic/Class.hs:737]: `-` and `negate` enter an infinite loop

File: symbolic-base/src/ZkFold/Base/Algebra/Basic/Class.hs [hunk 4]
```diff
@@ -690,3 +690,76 @@ instance FromConstant b a => FromConstant b (p -> a) where
...
+
+    negate :: Maybe a -> Maybe a
+    negate x = zero - x
+
+instance (Semiring a, AdditiveGroup a, FromConstant Integer a) => Ring (Maybe a)
+
+instance Field a => Field (Maybe a) where
+    (//) :: Maybe a -> Maybe a -> Maybe a
+    x // y = x * finv y
```
Comment (TurtlePU) [inline symbolic-base/src/ZkFold/Base/Algebra/Basic/Class.hs:746]: `//` and `finv` enter an infinite loop

File: symbolic-base/src/ZkFold/Base/Algebra/Basic/Class.hs [hunk 5]
```diff
@@ -690,3 +690,76 @@ instance FromConstant b a => FromConstant b (p -> a) where
...
+
+    finv :: Maybe a -> Maybe a
+    finv x = one // x
+
+    rootOfUnity :: Natural -> Maybe (Maybe a)
```
Comment (TurtlePU) [inline symbolic-base/src/ZkFold/Base/Algebra/Basic/Class.hs:751]: It's better to wrap in `Just` the result of `rootOfUnity @a`

File: symbolic-base/src/ZkFold/Base/Algebra/Basic/Class.hs [hunk 6]
```diff
@@ -690,3 +690,76 @@ instance FromConstant b a => FromConstant b (p -> a) where
...
+    rootOfUnity 0 = Just one
+    rootOfUnity _ = Nothing
+
+
+instance ToConstant a => ToConstant (Maybe a) where
+    type Const (Maybe a) = Natural
+    toConstant :: Maybe a -> Const (Maybe a)
+    toConstant = toConstant
```
Comment (TurtlePU) [inline symbolic-base/src/ZkFold/Base/Algebra/Basic/Class.hs:759]: `toConstant` loops forever

File: symbolic-base/src/ZkFold/Base/Algebra/Basic/Class.hs [hunk 7]
```diff
@@ -690,3 +690,76 @@ instance FromConstant b a => FromConstant b (p -> a) where
...
+    finv :: Maybe a -> Maybe a
+    finv x = one // x
+
+    rootOfUnity :: Natural -> Maybe (Maybe a)
+    rootOfUnity 0 = Just one
+    rootOfUnity _ = Nothing
+
+
+instance ToConstant a => ToConstant (Maybe a) where
+    type Const (Maybe a) = Natural
```
Comment (TurtlePU) [inline symbolic-base/src/ZkFold/Base/Algebra/Basic/Class.hs:757]: It should be `Maybe (Const a)`

File: symbolic-base/src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Internal.hs [hunk 8]
```diff
@@ -190,27 +190,37 @@ instance
...
+            ConstVar cV -> fromConstant cV
+          return $ SysVar (NewVar v)
+
 
     constraint p =
       let
         evalConstVar = \case
           SysVar sysV -> var sysV
-          ConstVar cV -> fromConstant cV
+          ConstVar cV -> if cV == zero
```
Comment (TurtlePU) [inline symbolic-base/src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Internal.hs:212]: There's nothing wrong with nonzero constants inside constraint; I think you meant to check the result of `p evalConstVar` if it is zero

File: symbolic-base/src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Internal.hs [hunk 9]
```diff
@@ -190,27 +190,38 @@ instance
...
     constraint p =
       let
         evalConstVar = \case
           SysVar sysV -> var sysV
           ConstVar cV -> fromConstant cV
+        r = p evalConstVar
       in
-        zoom #acSystem . modify $ insert (toVar (p at)) (p evalConstVar)
+        if r == zero
+          then return $ error "constrained variable not equal zero"
```
Comment (TurtlePU) [inline symbolic-base/src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Internal.hs:216]: I think, if `r == zero`, then actually constraint is satisfied just fine :)

File: symbolic-base/src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Internal.hs [hunk 10]
```diff
@@ -212,7 +212,7 @@ instance
             SysVar sysV -> var sysV
             ConstVar cV -> fromConstant cV
           r = p evalConstVar
-      in if r == zero
+      in if null $ variables r
         then return ()
```
Comment (TurtlePU) [inline symbolic-base/src/ZkFold/Symbolic/Compiler/ArithmeticCircuit/Internal.hs:216]: Here we can branch on whether `r` us zero and throw error if it is not

**Review by TurtlePU — APPROVED** [review]

**Review by vlasin — APPROVED** [review]

**Pull request merged**